go-controller/pkg/ovn/policy.go

```diff
@@ -666,15 +666,42 @@ func (oc *Controller) processLocalPodSelectorSetPods(policy *knet.NetworkPolicy,
 		}
 
 		// Get the logical port info
+		// If the LSP is Scheduled for removal, wait until it's either gone
+		// or re-added to ensure we have the up to date version in the cache
+		// this usually only occurs with stateful-sets
 		logicalPort := util.GetLogicalPortName(pod.Namespace, pod.Name)
 		portInfo, err := oc.logicalPortCache.get(logicalPort)
 		// pod is not yet handled
 		// no big deal, we'll get the update when it is.
 		if err != nil {
+			klog.Warning("Failed to get LSP for pod %v/%v for networkPolicy %v err: %v", pod.Namespace, pod.Name, policy.Name, err)
 			continue
 		}
 
-		// this pod is somehow already added to this policy, then skip
+		for i := 0; i < 5; i++ {
+			if portInfo.expires.IsZero() {
+				klog.V(5).Infof("Fresh LSP %v for network policy %v found in cache", portInfo.name, policy.Name)
+				break
+			}
+			// 24 is chosen because gomega.Eventually default timeout is 50ms
+			// goovn transactions take less than 50ms usually as well so np create
+			// should be done within a couple iterations
+			time.Sleep(24 * time.Millisecond)
+
+			// Re-fetch the LSP
+			// This should not fail, since we already checked if pod was handled
+			portInfo, err = oc.logicalPortCache.get(logicalPort)
+			if err != nil {
+				klog.Errorf("Failed to get LSP for pod %v/%v for networkPolicy %v err: %v", pod.Namespace, pod.Name, policy.Name, err)
+			}
+		}
+
+		// Ensure the LSP is still not scheduled for deletion
+		if !portInfo.expires.IsZero() {
+			klog.Errorf("LSP %+v is stale and should not be added to the network policy %v", portInfo, policy.Name)
```

Comment:
Also, if it has been deleted / expired, shouldn't we ignore the item and `continue`?

Reply:
Yes good catch thanks!